src/Storage/WriteStream.php

```diff
@@ -0,0 +1,105 @@
+<?php
+/**
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Storage;
+
+use Google\Cloud\Upload\AbstractUploader;
+use GuzzleHttp\Psr7\StreamDecoratorTrait;
+use GuzzleHttp\Psr7\BufferStream;
+use Psr\Http\Message\StreamInterface;
+
+/**
+ * A Stream implementation that uploads in chunks to a provided uploader when
+ * we reach a certain chunkSize. Upon `close`, we will upload the remaining chunk.
+ */
+class WriteStream implements StreamInterface
+{
+    use StreamDecoratorTrait;
+
+    private $uploader;
+    private $stream;
+    private $chunkSize = 262144;
+    private $hasWritten = false;
+
+    /**
+     * Create a new WriteStream instance
+     *
+     * @param AbstractUploader $uploader The uploader to use.
+     * @param array $options [optional] {
+     *      Configuration options.
+     *
+     *      @type int $chunkSize The size of the buffer above which we attempt to
+     *            upload data
+     * }
+     */
+    public function __construct($uploader = null, $options = [])
```

Comment:
There's an unfortunate circular dependency here between the WriteStream and the StreamableUploader. The uploader requires the data stream at instantiation. We could either set the uploader on the stream after initiation or set the data on the uploader after initiation.

I wanted to separate the decision of when to upload from the process of uploading. The goal was to try and use the `ResumableUploader` here as well, but it just moved the ugliness into there (handling 3 upload cases in the single class).